servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java

```diff
@@ -68,128 +70,193 @@ public LRUEvictor(BlockMetadataManager meta) {
     }
   }
 
-  private boolean alreadyAvailable(long bytesToBeAvailable, BlockStoreLocation location)
+  /**
+   * @return a StorageDir in the range of location that already has availableBytes larger than
+   *         bytesToBeAvailable, otherwise null
+   */
+  private StorageDir alreadyAvailable(long bytesToBeAvailable, BlockStoreLocation location)
       throws IOException {
     if (location.equals(BlockStoreLocation.anyTier())) {
-      for (StorageTier tier : mMeta.getTiers()) {
+      for (StorageTier tier : mMetaManager.getTiers()) {
         for (StorageDir dir : tier.getStorageDirs()) {
           if (dir.getAvailableBytes() >= bytesToBeAvailable) {
-            return true;
+            return dir;
           }
         }
       }
-      return false;
+      return null;
     }
 
     int tierAlias = location.tierAlias();
-    StorageTier tier = mMeta.getTier(tierAlias);
+    StorageTier tier = mMetaManager.getTier(tierAlias);
     if (location.equals(BlockStoreLocation.anyDirInTier(tierAlias))) {
       for (StorageDir dir : tier.getStorageDirs()) {
         if (dir.getAvailableBytes() >= bytesToBeAvailable) {
-          return true;
+          return dir;
         }
       }
-      return false;
+      return null;
     }
 
     StorageDir dir = tier.getDir(location.dir());
-    return dir.getAvailableBytes() >= bytesToBeAvailable;
+    return (dir.getAvailableBytes() >= bytesToBeAvailable) ? dir : null;
   }
 
-  @Override
-  public EvictionPlan freeSpace(long bytesToBeAvailable, BlockStoreLocation location)
-      throws IOException {
-    List<Pair<Long, BlockStoreLocation>> toMove = new ArrayList<Pair<Long, BlockStoreLocation>>();
-    List<Long> toEvict = new ArrayList<Long>();
-    EvictionPlan plan = null;
+  /**
+   * A recursive implementation of cascading LRU eviction.
+   *
+   * It will try to free space in next tier to transfer blocks there, if the next tier does not have
+   * enough free space to hold the blocks, the next next tier will be tried and so on until the
+   * bottom tier is reached, if blocks can not even be transferred to the bottom tier, they will be
+   * evicted, otherwise, only blocks to be freed in the bottom tier will be evicted.
+   *
+   * this method is only used in {@link #freeSpace(long, tachyon.worker.block.BlockStoreLocation)}
+   *
+   * @param bytesToBeAvailable bytes to be available after eviction
+   * @param location target location to evict blocks from
+   * @param plan the plan to be recursively updated, is empty when first called in
+   *        {@link #freeSpace(long, tachyon.worker.block.BlockStoreLocation)}
+   * @return the first StorageDir in the range of location to evict/move bytes from, or null if
+   *         there is no plan
+   */
+  private StorageDir freeSpace(long bytesToBeAvailable, BlockStoreLocation location,
```

Comment:
Can this private method have a more descriptive name other than overloading `freeSpace`? Usually we don't have overloading methods unless two methods are really doing identical things except taking different parameters.